### ISSUE TITLE
Tool for segmenting long audio files with erroneous transcripts

### ISSUE THREAD

**Issue by vijayaditya** [kaldi-asr/kaldi#869]
We need a script for segmenting long audio files with possibly erroneous transcripts. This would be useful for creating recipes like `librispeech` with other openly available databases. 
(This was a feature request from ASR researchers working in other languages.)

This tool was discussed in a different issue ( #718). I am creating this issue to keep track of this project. I am reproducing @danpovey 's comments here for convenience.

`
RE: "If possible it would be good to have scripts used for preparing the
Librispeech recipe".
I don't think Vassil plans to release them as open source. However, I have
a general plan to have someone work on this problem in the next few months
and create something that we can use in this type of scenario. Probably
we'd do some kind of phone-level alignment based on a phone recognition to
get an initial candidate segmentation, and then apply Vimal's cleanup
scripts, that he is working on, to get rid of bad segments. This means
that we can re-use Vimal's data-cleanup work and will need minimal extra
work to handle the long-audio-segmentation problem.
`

**Comment by danpovey** [comment]
In the next few weeks I'll be working on Vimal's data-cleanup stuff, and
I'll keep this scenario in mind.
Dan

**Comment by sikoried** [comment]
For rough transcription, I had good results training a 3-gram LM just on
the transcript and decoding (not aligning) that with a rather narrow beam.

**Comment by vince62s** [comment]
After the fix https://github.com/kaldi-asr/kaldi/pull/851 (for bi-grams use) I must say that the run_segmentation.sh script works very well if done after 2 passes.
Pass 1: take a basic existing gmm model to extract a few data to build a new tri3 model
Pass 2: reextract the data, gives me a very good ratio.

**Comment by danpovey** [comment]
I do actually want to refactor the long-data-segmentation script to have a nicer interface and to use the newer utilities as used by steps/cleanup/clean_and_segment_data.sh, e.g. to use steps/cleanup/make_biased_lm_graphs.sh.

@xiaohui-zhang, I am hoping that you might have time to do this, as you're working with these types of things right now... if not, I may ask Vimal - or Vimal may want to do it himself.

Firstly, I want there to be a *single* script in steps/cleanup-- e.g. steps/cleanup/segment_long_utterances.sh-- that is responsible for this, i.e. I don't want the user to have to call a large number of scripts.
Let's assume that some kind of model is available.  We'll have the usage something like this:
```
  Usage: $0 [options] <model-dir> <data-in> <segmented-data-out> <work-dir>
  E.g.: $0 exp/tri3 data/train_unsegmented data/train_segmented exp/tri3_segmentation
  This script performs segmentation of the data in <data-in>, writing the segmented
  data (with a segments file) to <segmented-data-out>.  The purpose of this script is
  to divide up the input data (which may consist of long recordings such as television
  shows or audiobooks) into segments which are of manageable length for further
  processing, along with the portion of the transcript that seems to match each segment.
  The output data is not necessarily particularly clean; you are advised to 
  run steps/cleanup/clean_and_segment_data.sh on the output in order to further
  clean it and eliminate data where the transcript doesn't seem to match.
 ```

Some of the steps will be similar to `egs/wsj/s5/local/run_segmentation.sh`.
We'll
   - (1) arbitrarily divide the input utterances into slightly overlapping utterances of reasonable length [e.g. 30sec max], without looking at the audio.  To do this you will need to get the utterance durations by calling `steps/data/get_utt2dur.sh <data-in>`, and then generating a segments file with overlapping segments, probably using a mechanism similar to whatever is used in egs/wsj/s5/local/run_segmentation.sh, and then applying `utils/data/subsegment_data_dir.sh` to apply this segments file [c.f. clean_and_segment_data.sh for how to do this.]  Note: it doesn't matter at this point what 'text' file you use, it won't ever be used.  Let's call this arbitrary-segmented data-directory `<data-uniform-segments>` at this point.

Note: the existing script run_segmentation does the feature generation on the split-up data itself, but in our case because the script doesn't know the feature type, so it will be better to ensure that the user has generated features for the data in `<data-in>`, and to rely on `subsegment_data_dir.sh` to correctly sub-segment the features.  subsegment_data_dir.sh automatically creates sub-matrices of the input matrices using the newly created 'matrix-range' mechanism.

 - (2) generate graphs for the input transcripts.  Because we have the transcript at the level of each utterance in `<data-in>`, it will be easiest to call `make_biased_lm_graphs.sh` on the original data in `<data-in>`, where we have one (very long) piece of text for each utterance.  Then you'll use a suitable invocation of utils/apply_map.pl to create a HCLG.fst.scp file that is indexed by the utterances in `<data-uniform-segments>`.

 - (3) decode the data in `<data-uniform-segments>` with the provided model and the graphs generated above.

 - (4) splice together the ctm files.. I think the existing run_segmentation.sh does this somehow, have a look.

 - (5) the actual segmentation.  The basic process will be similar to what's done in steps/cleanup/make_segmentation_data_dir.sh, but the output will just be the 'segments' file [a segments file that 's relative to the old utterance, as done in clean_and_segment_data.sh] and the 'text' file; creating the actual data-dir will be done by `utils/data/subsegment_data_dir.sh`.
   There are a few differences though [and please rewrite it from scratch.]

     - Firstly, let's not do a Levenshtein alignment, let's do a Smith-Waterman alignment-- the difference being that we allow arbitrary deletions at the beginning and end of the reference transcript, so the data is allowed without penalty to consume only part of the 'text' file.  [this may be useful in certain situations.]  You could consider implementing it in python (put it in steps/cleanup/internals/) and calling it in parallel using `$cmd`.
     - We don't need to be as careful as `make_segmentation_data_dir.sh` about finding 'clean' cut points and having WER cutoffs to remove bad segments, because we assume that the user will be running clean_and_segment_data.sh afterwards.  

An alternative approach that could be tried is to compute the segmentation into reasonably long chunks from the ctm, [e.g. chunks of at least 20 words] before looking at the reference transcript, and to apply the Smith-Waterman algorithm on each reasonably long chunk.  This will make it so that if, say, chapters are in the wrong order, it won't affect things too badly.  Hopefully we'd later remove 'bad' data in `utils/data/clean_and_segment_data.sh.`.  Anyway, the underlying tools for this wouldn't be too different.

We also need a good test bed for this stuff.  Librispeech would probably be a good place to look for this- i.e. the original Librispeech audiobooks.  Perhaps getting the test-bed could be considered a separate issue from actually writing the scripts, and someone else could help with that.

**Comment by vimalmanohar** [comment]
I can work on this.

**Comment by vijayaditya** [comment]
@vimalmanohar If you are working on this can you create a WIP PR and link it here. It would be interesting to follow the work.
